packages/xod-client/src/editor/actions.js

```diff
@@ -710,39 +726,69 @@ export const runTabtest = patchPath => (dispatch, getState) => {
     ProjectSelectors.getProject
   )(getState());
 
+  const ABORT_ERROR_TYPE = 'ABORT_BY_USER';
+  const shouldContinue = () => Selectors.isTabtestRunning(getState());
+  const abortOrPass = fn => x => {
+    if (!shouldContinue()) {
+      return Promise.reject(createError(ABORT_ERROR_TYPE, {}));
+    }
+    return fn(x);
+  };
+
   suiteP
     .then(
-      R.tap(() => {
-        dispatch({ type: ActionType.TABTEST_GENERATED_CPP });
-      })
+      abortOrPass(
+        R.tap(() => {
+          dispatch({ type: ActionType.TABTEST_GENERATED_CPP });
+        })
+      )
+    )
+    .then(abortOrPass(XCT.compileSuite(HOSTNAME)))
+    .then(
+      abortOrPass(
+        R.tap(() => {
+          dispatch({ type: ActionType.TABTEST_COMPILED });
+        })
+      )
     )
-    .then(compileSuite(HOSTNAME))
     .then(
-      R.tap(() => {
-        dispatch({ type: ActionType.TABTEST_COMPILED });
+      abortOrPass(suite =>
+        runWasmWorker(suite, worker =>
+          dispatch({
+            type: ActionType.TABTEST_LAUNCHED,
+            payload: { worker },
+          })
+        )
+      )
+    )
+    .then(
+      abortOrPass(({ stdout, worker }) => {
```

Comment:
These repetitive `.then(abortOrPass(...))` is yet another reason to upgrade to latest Ramda. Its `R.composeWith` and `R.pipeWith` make this chain more elegant and shallowly nested.

Just a note, not a change request.